File: tooling/gx-chat-core/extras-parity.ts
/*
 * Family f's parity fixture: runs the shipped TypeScript rules over a table of inputs and writes
 * what they produced, so `packages/gx-chat-core/examples/extras_parity.rs` can run the Rust port
 * over the same table and diff the two.
 *
 * The replay gate (`docs/2026-09-21/rust-chat/REPLAY.md`) drives the whole brain, which needs
 * family a's fold to be in place first. This covers family f's own rules in the meantime, and it
 * keeps covering the branches a recording never reaches: a stale fleet roster, a blocked task, a
 * rule-only terminal screen, a reserved file name.
 *
 * Everything in the table is invented. It carries no session, project or conversation data, so the
 * output is safe to commit and safe to read.
 *
 * Run: bun tooling/gx-chat-core/extras-parity.ts [output.json]
 */

import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname } from 'node:path';

import { sessionChatAgentFleetRows, subagentModelLabel } from '@/packages/shared/session-chat-presentation/agent-fleet';
import { sessionChatAgentTaskPanel } from '@/packages/shared/session-chat-presentation/agent-tasks';
import {
  formatSessionChatActivityElapsed,
  sessionChatActivityElapsedSeconds,
} from '@/packages/shared/session-chat-controller/activity';
import { formatSessionTerminalTailPreview } from '@/packages/shared/session-chat-presentation/terminal-tail';
import {
  SESSION_CHAT_MINIMAP,
  sessionChatMinimapPreview,
  sessionChatMinimapPreviewText,
} from '@/packages/shared/session-chat-presentation/minimap';
import {
  sessionChatSearchCountLabel,
  sessionChatTranscriptItemText,
  sessionChatTranscriptMatches,
} from '@/packages/shared/session-chat-presentation/transcript-search';
import {
  isSessionChatSubagentSelf,
  sessionChatToolSubagent,
} from '@/packages/shared/session-chat-presentation/subagent';
import {
  sessionChatNewSessionWelcomeTitle,
  sessionChatShowsNewSessionWelcome,
  sessionChatWelcomeAgentIcon,
  sessionChatWelcomeAgentName,
} from '@/packages/shared/session-chat-presentation/new-session-welcome';
import { sessionChatEmptyStateCopy } from '@/packages/core-ui/chat/session-chat-empty-state';
import { formatSessionChatDuration } from '@/packages/core-ui/chat/session-chat-duration';
import {
  folderPathError,
  markdownStemError,
  normalizedFolderPath,
  normalizedMarkdownStem,
  suggestedMarkdownStem,
} from '@/packages/shared/session-chat-presentation/save-markdown';
import { SESSION_CHAT_WORKING_WORDS } from '@/packages/shared/session-chat-presentation/working-words';

/** A fixed instant, so the elapsed clocks are the same on every machine. */
const NOW = Date.parse('2026-09-22T12:00:00.000Z');

const cases: Record<string, { input: unknown; output: unknown }[]> = {};
function record(group: string, input: unknown, output: unknown): void {
  (cases[group] ??= []).push({ input, output });
}

// --- working words ---------------------------------------------------------
for (const draw of [0, 0.04182539903558791, 0.5, 0.723172509809956, 0.9999999]) {
  record(
    'workingWord',
    draw,
    SESSION_CHAT_WORKING_WORDS[Math.floor(draw * SESSION_CHAT_WORKING_WORDS.length)] ?? 'Working'
  );
}

// --- activity clock --------------------------------------------------------
for (const seconds of [0, 0.4, 1, 59, 60, 61, 3599, 3600, 3661, 86_399]) {
  record('activityElapsed', seconds, formatSessionChatActivityElapsed(seconds));
}
for (const input of [
  { elapsedSeconds: 12, detectedAt: '2026-09-22T11:59:30.000Z' },
  { elapsedSeconds: 12, detectedAt: '2026-09-22T12:00:30.000Z' },
  { elapsedSeconds: 0, detectedAt: 'not a date' },
  { detectedAt: '2026-09-22T11:59:30.000Z' },
]) {
  record('activitySeconds', input, sessionChatActivityElapsedSeconds(input, NOW));
}

// --- model and effort labels ----------------------------------------------
for (const info of [
  {},
  { model: 'claude-opus-4-5' },
  { model: 'claude-opus-4-5', effort: 'xhigh' },
  { model: 'Opus 5 [1m]', effort: 'high' },
  { model: 'claude-3-5-sonnet-20241022' },
  { model: 'gpt-5.1-astra' },
  { model: 'gpt-5-codex', effort: 'medium' },
  { model: 'some-other-model' },
  { model: '  haiku  ' },
]) {
  record('modelLabel', info, subagentModelLabel(info));
}

// --- the fleet strip -------------------------------------------------------
const fleets: { fleet: unknown; provider: string | null }[] = [
  { fleet: null, provider: 'claude' },
  { fleet: { agents: [], detectedAt: '2026-09-22T11:59:00.000Z' }, provider: 'claude' },
  {
    fleet: {
      detectedAt: '2026-09-22T11:59:00.000Z',
      agents: [
        {
          id: 'a1',
          name: 'general-purpose',
          task: 'Audit the config',
          status: 'working',
          elapsedSeconds: 30,
          tokens: '↓ 12.1k tokens',
          model: 'claude-opus-4-5',
          effort: 'high',
        },
        { name: 'explorer', status: 'idle', nested: 2 },
      ],
    },
    provider: 'claude',
  },
  {
    fleet: {
      detectedAt: '2026-09-22T11:59:00.000Z',
      stale: true,
      agents: [{ name: 'worker', task: 'x', status: 'working', elapsedSeconds: 5 }],
    },
    provider: 'codex',
  },
  {
    fleet: {
      detectedAt: '2026-09-22T11:59:00.000Z',
      validUntil: '2026-09-22T11:59:30.000Z',
      agents: [{ name: 'worker', status: 'working' }],
    },
    provider: 'codex',
  },
  {
    fleet: {
      detectedAt: '2026-09-22T11:59:00.000Z',
      validUntil: 'nonsense',
      agents: [{ name: 'worker', status: 'working', elapsedSeconds: 7 }],
    },
    provider: null,
  },
];
for (const { fleet, provider } of fleets) {
  record('fleetStrip', { fleet, provider, now: NOW }, sessionChatAgentFleetRows(fleet as never, provider, NOW));
}

// --- the task panel --------------------------------------------------------
const tasks = {
  tasks: [
    { id: '3', subject: 'Write the tests', status: 'pending', blockedBy: ['1', '2', '9'] },
    { id: '1', subject: 'Read the code', status: 'completed' },
    { id: '2', subject: 'Draft the change', activeForm: 'Drafting the change', status: 'in_progress' },
    { id: '10', subject: 'Ship it', status: 'weird-status' },
    { id: 'x', subject: 'Unnumbered', status: 'completed' },
  ],
};
for (const collapsed of [true, false]) {
  for (const showCompleted of [true, false]) {
    record(
      'taskPanel',
      { tasks, collapsed, showCompleted },
      sessionChatAgentTaskPanel(tasks as never, { collapsed, showCompleted })
    );
  }
}
record(
  'taskPanel',
  { tasks: null, collapsed: false, showCompleted: false },
  sessionChatAgentTaskPanel(null, { collapsed: false, showCompleted: false })
);
record(
  'taskPanel',
  { tasks: { tasks: [] }, collapsed: false, showCompleted: false },
  sessionChatAgentTaskPanel({ tasks: [] }, { collapsed: false, showCompleted: false })
);

// --- the terminal tail preview --------------------------------------------
for (const lines of [
  [],
  ['  hello', ''],
  ['────────────────────────────────────────────', '  a short line', '────────────────────────────────────────────'],
  ['━━━━━━━━', 'x'],
  ['-------', 'seven dashes are not a rule'],
  ['a──────b', 'mixed'],
]) {
  record('tailPreview', lines, formatSessionTerminalTailPreview(lines));
}

// --- the minimap -----------------------------------------------------------
const longText = 'word '.repeat(60);
for (const message of [
  null,
  { id: 'm1', blocks: [] },
  {
    id: 'm2',
    blocks: [
      { type: 'text', text: '  spaced \n out  ' },
      { type: 'tool-call', name: 'x' },
    ],
  },
  { id: 'm3', blocks: [{ type: 'text', text: longText }] },
]) {
  record('minimapPreview', message, {
    text: sessionChatMinimapPreviewText(message as never),
    preview: sessionChatMinimapPreview(message as never),
  });
}
record('minimapGeometry', null, SESSION_CHAT_MINIMAP);

// --- transcript search -----------------------------------------------------
const items = [
  { kind: 'message', message: { id: 'm1', text: 'the cat sat on the mat' } },
  {
    kind: 'summary',
    id: 't1',
    user: { id: 'u1', text: 'cat' },
    final: { id: 'f1', text: 'CAT cat' },
    work: [{ id: 'w1', tools: [{ call: { name: 'Cat' }, preview: 'cat preview' }], files: [{ path: '/tmp/cat.txt' }] }],
  },
  {
    kind: 'completed-work',
    id: 't2',
    label: 'Worked for 3s',
    artifacts: [{ id: 'a1', suppressed: { label: 'a cat card' } }],
  },
];
for (const query of ['', '  ', 'cat', 'CAT', 'nothing']) {
  record('searchMatches', query, sessionChatTranscriptMatches(items, query));
  record(
    'searchLabel',
    { query, total: sessionChatTranscriptMatches(items, query).length, activeIndex: 0 },
    sessionChatSearchCountLabel(query, sessionChatTranscriptMatches(items, query).length, 0)
  );
}
for (const item of items) {
  record('searchItemText', item, sessionChatTranscriptItemText(item));
}

// --- the subagent target ---------------------------------------------------
const targets: { call: unknown; result: unknown; agentPath?: string }[] = [
  { call: { type: 'tool-call', name: 'Read', input: {} }, result: undefined },
  {
    call: { type: 'tool-call', name: 'Task', input: { description: 'Audit', subagent_type: 'general-purpose' } },
    result: { type: 'tool-result', output: 'agentId: abc_123 done' },
  },
  {
    call: { type: 'tool-call', name: 'mcp.spawn_agent', input: { task_name: 'audit' } },
    result: { type: 'tool-result', output: '{"agent_nickname":"Auditor"}' },
  },
  {
    call: { type: 'tool-call', name: 'send_message', input: { target: 'child' } },
    result: undefined,
    agentPath: '/root/parent',
  },
  { call: { type: 'tool-call', name: 'send_message', input: { id: '/abs/child' } }, result: undefined },
  { call: { type: 'tool-call', name: 'followup_task', input: {} }, result: undefined },
];
for (const { call, result, agentPath } of targets) {
  record(
    'subagentTarget',
    { call, result, agentPath: agentPath ?? '/root' },
    sessionChatToolSubagent(call as never, result as never, agentPath)
  );
}
for (const selector of ['/root', '/root/parent', '/other']) {
  record('subagentSelf', { selector, agentPath: '/root/parent' }, isSessionChatSubagentSelf(selector, '/root/parent'));
}

// --- the welcome and the empty state --------------------------------------
for (const kind of ['loading', 'empty', 'error', 'unsupported', 'starting', 'ready', 'notFound']) {
  record('showsWelcome', kind, sessionChatShowsNewSessionWelcome(kind));
}
for (const label of [null, '', '  ', 'codex', 'claude', 'claude-code', 'my_custom-agent', 'hermes', 'grok', 'omp']) {
  const name = sessionChatWelcomeAgentName(label);
  record('welcomeAgentName', label, name);
  record('welcomeTitle', name, sessionChatNewSessionWelcomeTitle(name));
  record('welcomeIcon', { label, icon: null }, sessionChatWelcomeAgentIcon(label, null) ?? null);
}
record('welcomeIcon', { label: 'codex', icon: 'browser' }, sessionChatWelcomeAgentIcon('codex', 'browser') ?? null);
record(
  'welcomeIcon',
  { label: 'codex', icon: 'not-an-icon' },
  sessionChatWelcomeAgentIcon('codex', 'not-an-icon') ?? null
);
for (const kind of ['loading', 'empty', 'error', 'unsupported', 'starting'] as const) {
  for (const agent of [null, 'codex', '  ']) {
    record('emptyState', { kind, agent }, sessionChatEmptyStateCopy(kind, agent));
  }
}

// --- durations -------------------------------------------------------------
for (const ms of [0, -5, 499, 500, 1_000, 59_400, 60_000, 3_600_000, 3_723_000]) {
  record('duration', ms, formatSessionChatDuration(ms));
}

// --- Save to Markdown paths ------------------------------------------------
for (const value of ['', '  ', 'notes', ' notes / deep ', 'a/./b', 'con', 'a<b', 'trailing.', 'x'.repeat(250)]) {
  record('folderError', value, folderPathError(value) ?? null);
  record('folderPath', value, normalizedFolderPath(value));
}
for (const value of ['', '  ', 'notes', 'notes.MD', 'notes.md ', 'nul.txt', 'a/b', '..', 'ends.', 'x'.repeat(130)]) {
  record('stemError', value, markdownStemError(value) ?? null);
  record('stem', value, normalizedMarkdownStem(value));
}
const existing = [
  'docs/2026-09-22/Fixing the build 1.md',
  'docs/2026-09-22/Fixing the build 4.md',
  'docs/2026-09-22/fixing the build 9.md',
  'docs/2026-09-22/nested/Fixing the build 99.md',
  'docs/other/Fixing the build 50.md',
  'docs/2026-09-22/Fixing the build.md',
];
for (const title of ['Fixing the build', 'Fix-the-build', '', '  ??  ', 'A'.repeat(140), 'Trailing dots...']) {
  record(
    'suggestedStem',
    { title, folder: '2026-09-22', existing },
    suggestedMarkdownStem(title, '2026-09-22', existing)
  );
}

const target = process.argv[2] ?? '/tmp/gx-chat/extras-parity.json';
mkdirSync(dirname(target), { recursive: true });
writeFileSync(target, `${JSON.stringify({ v: 1, now: NOW, cases }, null, 2)}\n`);
let total = 0;
for (const group of Object.values(cases)) total += group.length;
console.log(`extras parity  ${Object.keys(cases).length} groups, ${total} cases`);
console.log(`wrote          ${target}`);

File: packages/gx-chat-core/examples/extras_parity.rs
//! Family f's parity gate: runs the Rust port over the same table
//! `tooling/gx-chat-core/extras-parity.ts` ran the shipped TypeScript over, and reports every case
//! whose answer differs.
//!
//! The replay gate (`docs/2026-09-21/rust-chat/REPLAY.md`) drives the whole brain and therefore
//! waits on family a's fold. This one covers family f's own rules on their own, including the
//! branches a recording never reaches.
//!
//! ```text
//! bun tooling/gx-chat-core/extras-parity.ts
//! cargo run --example extras_parity
//! ```
//!
//! The fixture is invented, so this example may print its values; a replay recording may not.

use std::collections::BTreeMap;
use std::process::ExitCode;

use ghostex_gx_chat_core::extras::activity::{activity_elapsed_seconds, format_activity_elapsed};
use ghostex_gx_chat_core::extras::agent_fleet::{agent_fleet_rows, subagent_model_label};
use ghostex_gx_chat_core::extras::agent_tasks::agent_task_panel;
use ghostex_gx_chat_core::extras::minimap_rail::{geometry, minimap_preview, minimap_preview_text};
use ghostex_gx_chat_core::extras::save_markdown_paths::{
    folder_path_error, markdown_stem_error, normalized_folder_path, normalized_markdown_stem,
    suggested_markdown_stem,
};
use ghostex_gx_chat_core::extras::subagent_target::{is_subagent_self, tool_subagent};
use ghostex_gx_chat_core::extras::terminal_tail_format::format_terminal_tail_preview;
use ghostex_gx_chat_core::extras::transcript_search::{
    search_count_label, transcript_item_text, transcript_matches,
};
use ghostex_gx_chat_core::extras::welcome::{
    empty_state_copy, format_duration, new_session_welcome_title, shows_new_session_welcome,
    welcome_agent_icon, welcome_agent_name,
};
use ghostex_gx_chat_core::extras::working_words::pick_working_word;
use ghostex_gx_chat_core::ChatContext;
use serde_json::{json, Value};

fn main() -> ExitCode {
    let path = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "/tmp/gx-chat/extras-parity.json".to_string());
    let Ok(text) = std::fs::read_to_string(&path) else {
        eprintln!("fixture not found: {path}");
        eprintln!("run: bun tooling/gx-chat-core/extras-parity.ts");
        return ExitCode::FAILURE;
    };
    let fixture: Value = serde_json::from_str(&text).expect("the fixture is JSON");
    let now = fixture["now"].as_f64().expect("the fixture carries `now`");
    let context = ChatContext::at(now);

    let mut checked = 0usize;
    let mut differences: BTreeMap<String, usize> = BTreeMap::new();
    let mut first: Vec<String> = Vec::new();
    let groups = fixture["cases"].as_object().expect("`cases` is an object");
    for (group, list) in groups {
        for (index, case) in list
            .as_array()
            .expect("a group is a list")
            .iter()
            .enumerate()
        {
            let input = &case["input"];
            let expected = &case["output"];
            let Some(actual) = run(group, input, &context) else {
                *differences
                    .entry(format!("{group} (not covered)"))
                    .or_default() += 1;
                continue;
            };
            checked += 1;
            if &actual != expected {
                *differences.entry(group.clone()).or_default() += 1;
                if first.len() < 12 {
                    first.push(format!(
                        "  {group}[{index}]\n    input    {input}\n    expected {expected}\n    actual   {actual}"
                    ));
                }
            }
        }
    }

    let total: usize = differences.values().sum();
    println!("fixture     {path}");
    println!("checked     {checked} cases in {} groups", groups.len());
    println!("differences {total}");
    for (group, count) in &differences {
        println!("  {group}  x{count}");
    }
    for line in &first {
        println!("{line}");
    }
    if total == 0 {
        ExitCode::SUCCESS
    } else {
        ExitCode::FAILURE
    }
}

/// The Rust answer for one case, or `None` when this build does not cover the group.
fn run(group: &str, input: &Value, context: &ChatContext) -> Option<Value> {
    Some(match group {
        "workingWord" => json!(pick_working_word(input.as_f64()?)),
        "activityElapsed" => json!(format_activity_elapsed(input.as_f64()?)),
        "activitySeconds" => number(activity_elapsed_seconds(
            input.get("elapsedSeconds").and_then(Value::as_f64),
            input.get("detectedAt").and_then(Value::as_str)?,
            context.now_ms,
            context.utc_offset_minutes,
        )),
        "modelLabel" => json!(subagent_model_label(
            input.get("model").and_then(Value::as_str),
            input.get("effort").and_then(Value::as_str),
        )),
        "fleetStrip" => {
            let fleet = input.get("fleet").filter(|value| !value.is_null());
            match agent_fleet_rows(fleet, input["provider"].as_str(), context) {
                Some(strip) => serde_json::to_value(strip).ok()?,
                None => Value::Null,
            }
        }
        "taskPanel" => {
            let tasks = input.get("tasks").filter(|value| !value.is_null());
            match agent_task_panel(
                tasks,
                input["collapsed"].as_bool()?,
                input["showCompleted"].as_bool()?,
            ) {
                Some(panel) => serde_json::to_value(panel).ok()?,
                None => Value::Null,
            }
        }
        "tailPreview" => {
            let lines: Vec<String> = input
                .as_array()?
                .iter()
                .map(|line| line.as_str().unwrap_or_default().to_string())
                .collect();
            json!(format_terminal_tail_preview(&lines))
        }
        "minimapPreview" => {
            let message = input.as_object().map(|_| input);
            json!({
                "text": minimap_preview_text(message),
                "preview": minimap_preview(message, geometry().preview_limit),
            })
        }
        "minimapGeometry" => serde_json::to_value(geometry()).ok()?,
        "searchMatches" => {
            serde_json::to_value(transcript_matches(&search_items(), input.as_str()?)).ok()?
        }
        "searchLabel" => json!(search_count_label(
            input["query"].as_str()?,
            input["total"].as_u64()? as usize,
            input["activeIndex"].as_u64()? as usize,
        )),
        "searchItemText" => json!(transcript_item_text(input)),
        "subagentTarget" => {
            let call = input.get("call").filter(|value| !value.is_null());
            let result = input.get("result").filter(|value| !value.is_null());
            match tool_subagent(call, result, input["agentPath"].as_str()?) {
                Some(target) => serde_json::to_value(target).ok()?,
                None => Value::Null,
            }
        }
        "subagentSelf" => json!(is_subagent_self(
            input["selector"].as_str()?,
            input["agentPath"].as_str()?
        )),
        "showsWelcome" => json!(shows_new_session_welcome(input.as_str()?)),
        "welcomeAgentName" => match welcome_agent_name(input.as_str()) {
            Some(name) => json!(name),
            None => Value::Null,
        },
        "welcomeTitle" => json!(new_session_welcome_title(input.as_str())),
        "welcomeIcon" => match welcome_agent_icon(
            input.get("label").and_then(Value::as_str),
            input.get("icon").and_then(Value::as_str),
        ) {
            Some(icon) => json!(icon),
            None => Value::Null,
        },
        "emptyState" => serde_json::to_value(empty_state_copy(
            input["kind"].as_str()?,
            input.get("agent").and_then(Value::as_str),
        ))
        .ok()?,
        "duration" => json!(format_duration(input.as_f64()?)),
        "folderError" => match folder_path_error(input.as_str()?) {
            Some(error) => json!(error),
            None => Value::Null,
        },
        "folderPath" => json!(normalized_folder_path(input.as_str()?)),
        "stemError" => match markdown_stem_error(input.as_str()?) {
            Some(error) => json!(error),
            None => Value::Null,
        },
        "stem" => json!(normalized_markdown_stem(input.as_str()?)),
        "suggestedStem" => {
            let existing: Vec<String> = input["existing"]
                .as_array()?
                .iter()
                .map(|path| path.as_str().unwrap_or_default().to_string())
                .collect();
            json!(suggested_markdown_stem(
                input["title"].as_str()?,
                input["folder"].as_str()?,
                &existing
            ))
        }
        _ => return None,
    })
}

/// A JavaScript number, or `null`: an integral value serializes without a fractional part.
fn number(value: Option<f64>) -> Value {
    match value {
        Some(value) if value.fract() == 0.0 => Value::from(value as i64),
        Some(value) => Value::from(value),
        None => Value::Null,
    }
}

/// The transcript rows the search cases run over, the same list the fixture used.
fn search_items() -> Vec<Value> {
    vec![
        json!({ "kind": "message", "message": { "id": "m1", "text": "the cat sat on the mat" } }),
        json!({
            "kind": "summary",
            "id": "t1",
            "user": { "id": "u1", "text": "cat" },
            "final": { "id": "f1", "text": "CAT cat" },
            "work": [{
                "id": "w1",
                "tools": [{ "call": { "name": "Cat" }, "preview": "cat preview" }],
                "files": [{ "path": "/tmp/cat.txt" }]
            }]
        }),
        json!({
            "kind": "completed-work",
            "id": "t2",
            "label": "Worked for 3s",
            "artifacts": [{ "id": "a1", "suppressed": { "label": "a cat card" } }]
        }),
    ]
}
